pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+// that numRegistered <= numAllowed.  This method will return true if we can
+// keep the invariant and added one to the number registered, otherwise it will
+// return false.
+func (r *RaftTracer) maybeRegister() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegistered := r.numRegistered.Load()
+
+	// The maximum number of traces has been reached. We don't register this
+	// trace and return false.
+	if numRegistered == numAllowed {
+		return false
+	}
+
+	// This can only happen if numAllowed has changed. If this happens flush all
+	// our current traces and don't register this request.
+	for numRegistered > numAllowed {
+		r.FlushAll()
+		return false
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	return r.numRegistered.CompareAndSwap(numRegistered, numRegistered+1)
+}
+
+// RegisterRemote is used to register a remote trace. This is used when we
+// receive a raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(e kvserverpb.TracedEntry) {
+	if !r.maybeRegister() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	ctx := logtags.AddTag(r.ctx, "id", tracingID(e.TraceID, e.SpanID))
+	t := traceValue{traceID: e.TraceID, spanID: e.SpanID, ctx: ctx, baseCtx: nil}
+	t.logf(1, "start trace for index %d from remote", e.Index)
+	r.m.Store(e.Index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if !r.maybeRegister() {
+		return
+	}
+	span := tracing.SpanFromContext(baseCtx.Context())
+	ctx := logtags.AddTag(r.ctx, "id", tracingID(span.TraceID(), span.SpanID()))
+	// NB: We grab the trace and span id now because the baseCtx.Context may be
+	// nil'ed later.
+	t := traceValue{traceID: span.TraceID(), spanID: span.SpanID(), ctx: ctx, baseCtx: baseCtx}
+
+	// Log the identical entry to both to make it easier to match up later.
+	t.logf(1, "registering entry index %d with tracing id (%s)", ent.Index, tracingID(t.traceID, t.spanID))
+	r.m.Store(kvpb.RaftIndex(ent.Index), &t)
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []kvserverpb.TracedEntry {
+	// NB: This check is an optimization to handle the common case where there
+	// are no registered traces on the store.
+	if r.numRegistered.Load() == 0 {
```

Comment:
Ideally, we should be checking a local variable here, instead of the global count. Better to avoid global synchronization.

Reply:
I'm not sure if there is a way to do this with a local variable. We could store a "replica level count" - but in practice this probably isn't any better. I can look at the performance of this, but I don't expect this will show up since it is only one extra main memory access per message.

Reply:
Having a range-local var is helpful for other reasons: see another comment in `maybeRegister` and `FlushAll` calls.